types/npm-definition.json

```diff
@@ -4,29 +4,205 @@
+    }
+  },
+  "common_patterns": [
+    {
+      "pattern": "pkg:npm/{name}@{version}",
+      "description": "Standard unscoped package",
+      "example": "pkg:npm/lodash@4.17.21"
+    },
+    {
+      "pattern": "pkg:npm/%40{scope}/{name}@{version}",
+      "description": "Standard scoped package with encoded scope",
+      "example": "pkg:npm/%40babel/core@7.22.0"
+    }
+  ],
+  "anti_patterns": [
+    "pkg:npm/@scope/package",
+    "pkg:npm/Package@1.0.0",
+    "pkg:npm/package@",
+    "pkg:npm/",
+    "pkg:npm"
+  ],
+  "parsing_decision_tree": {
+    "steps": [
+      {
+        "question": "Does the package have a scope?",
+        "yes": "Look for %40 at the beginning of the name portion",
+        "no": "Proceed with unscoped package parsing"
+      },
+      {
+        "question": "Is there a version component?",
+        "yes": "Parse everything after the last @ as version",
+        "no": "Version is optional"
+      },
+      {
+        "question": "Are there qualifiers?",
+        "yes": "Parse key=value pairs after ?",
+        "no": "No qualifiers present"
+      }
+    ]
+  },
+  "context": {
+    "ecosystem": "Node.js",
+    "package_manager": "npm",
+    "url_scheme": "pkg",
+    "hierarchical_structure": "pkg:npm/[scope/]{name}@{version}?{qualifiers}",
+    "backward_compatibility": "Supports legacy case-sensitive names for old packages",
+    "forward_compatibility": "New packages must follow strict lowercase rules"
+  },
+  "edge_cases": [
+    {
+      "case": "Legacy packages with uppercase names",
+      "handling": "While new packages must be lowercase, existing packages in registries may have mixed case",
+      "purl_impact": "PURLs should normalize to lowercase for consistency"
+    },
+    {
+      "case": "Scoped packages with dots/underscores",
+      "handling": "Scoped packages can start with dots/underscores, unscoped cannot",
+      "purl_impact": "Namespace validation differs from name validation"
+    }
+  ],
+  "llm_guidance": {
+    "do_not_assume": [
+      "All npm packages have scopes",
+      "Package names can have any characters",
+      "Versions don't need to follow SemVer",
+      "@ symbols don't need encoding",
+      "Case doesn't matter for package names"
+    ],
+    "key_facts_to_remember": [
+      "Scope @ must be encoded as %40",
+      "New package names must be lowercase",
+      "Names must be URL-safe",
+      "214 character limit includes scope",
+      "Scoped packages can start with dots/underscores"
+    ],
+    "common_mistakes_to_avoid": [
+      "Forgetting to encode the scope @ symbol",
+      "Using uppercase in new package names",
+      "Including spaces without encoding",
+      "Misunderstanding scope vs name rules"
+    ]
+  },
   "examples": [
-    "pkg:npm/foobar@12.3.1",
-    "pkg:npm/%40angular/animation@12.3.1",
-    "pkg:npm/mypackage@12.4.5?vcs_url=git://host.com/path/to/repo.git%404345abcd34343"
-  ]
+    {
```

Comment:
I guess these would be tests instead?

Reply:
Currently, it is a list. For a machine to learn, we need to explain why a given example is correct or incorrect using a description attribute. Otherwise, it will assume that only those tokens are correct.